model/util.go

```diff
@@ -17,114 +17,256 @@ package model
+	}
+
+	data = bytes.TrimSpace(data)
+	if data[0] == '{' && parameterName != "constants" && parameterName != "timeouts" {
+		extractData := map[string]json.RawMessage{}
+		err = json.Unmarshal(data, &extractData)
+		if err != nil {
+			return &UnmarshalError{
+				err:           err,
+				parameterName: parameterName,
+				primitiveType: reflect.TypeOf(*valObject).Kind(),
+			}
+		}
+
+		var ok bool
+		if data, ok = extractData[parameterName]; !ok {
+			return fmt.Errorf("external resource parameter not found: %q", parameterName)
+		}
+	}
+
+	return unmarshalObject(parameterName, data, valObject)
 }
 
-func primitiveOrStruct[T any, U any](data []byte) (valStruct *U, valPrimitive T, err error) {
-	if data[0] == '{' {
-		err = json.Unmarshal(data, &valStruct)
+func unmarshalPrimitiveOrObject[T string | bool, U any](parameterName string, data []byte, valPrimitive *T, valStruct *U) error {
+	data = bytes.TrimSpace(data)
+	if len(data) == 0 {
+		// TODO: Normalize error messages
+		return fmt.Errorf("%s no bytes to unmarshal", parameterName)
+	}
+
+	isObject := data[0] == '{' || data[0] == '['
+	var err error
+	if isObject {
+		err = unmarshalObject(parameterName, data, valStruct)
 	} else {
-		err = json.Unmarshal(data, &valPrimitive)
+		err = unmarshalPrimitive(parameterName, data, valPrimitive)
+	}
+
+	var unmarshalError *UnmarshalError
+	if errors.As(err, &unmarshalError) {
+		unmarshalError.objectType = reflect.TypeOf(*valStruct).Kind()
+		unmarshalError.primitiveType = reflect.TypeOf(*valPrimitive).Kind()
 	}
-	return
+
+	return err
 }
 
-func unmarshalKey(key string, data map[string]json.RawMessage, output interface{}) error {
-	if _, found := data[key]; found {
-		if err := json.Unmarshal(data[key], output); err != nil {
-			return fmt.Errorf("failed to  unmarshall key '%s' with data'%s'", key, data[key])
+func unmarshalPrimitive[T string | bool](parameterName string, data []byte, value *T) error {
+	if value == nil {
+		return nil
+	}
+
+	err := json.Unmarshal(data, value)
+	if err != nil {
+		return &UnmarshalError{
+			err:           err,
+			parameterName: parameterName,
+			primitiveType: reflect.TypeOf(*value).Kind(),
 		}
 	}
+
 	return nil
 }
 
-// unmarshalFile same as calling unmarshalString following by getBytesFromFile.
-// Assumes that the value inside `data` is a path to a known location.
-// Returns the content of the file or a not nil error reference.
-func unmarshalFile(data []byte) (b []byte, err error) {
-	filePath, err := unmarshalString(data)
-	if err != nil {
-		return nil, err
+func unmarshalObject[U any](parameterName string, data []byte, value *U) error {
+	if value == nil {
+		return nil
 	}
-	file, err := getBytesFromFile(filePath)
+
+	// Removed to maintain the golang 1.19 compatibility
+	// just define another type to unmarshal object, so the UnmarshalJSON will not be called recursively
+	// type forUnmarshal *U
+	// valueForUnmarshal := new(forUnmarshal)
+	// *valueForUnmarshal = value
```

Comment:
Won't this be a problem? If not, we can remove this comment since it can be confusing for later maintainers.